# **Exploring Ebay Car Sales Data**
This project aims to clean and analyse a [dataset](https://www.kaggle.com/orgesleka/used-cars-database/data) containing information about car listing on Ebay. 
The dataset contains a number of variables including:
- `Name of the car`
- `Seller type (private or dealer`
- `Price`
- `Vehicle type`
- `Car brand`
- `Fuel type`
- `Kilometres the car has driven`
- `Number of pictures contained within the advert`
- `Date the advert was created`

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [5]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In the cells above we imported the dataset as a panda dataframe and inspected it. From this we can see can observe the following:
- It has 371528 rows and 20 columns
- Of those columns, 13 are objects and 7 are integers
- 5 columns contain null values

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
columns = autos.columns
cols_dict = {'dateCrawled': 'date_crawled',
'name': 'name', 
'seller': 'seller',
'offerType': 'offer_type', 
'price': 'price', 
'abtest': 'abtest',
'vehicleType': 'vehicle_type', 
'yearOfRegistration': 'registration_year', 
'gearbox': 'gearbox', 
'powerPS': 'power_ps', 
'model': 'model',
'kilometer': 'kilometer', 
'monthOfRegistration': 'registration_month',
'fuelType': 'fuel_type',
'brand': 'brand',
'notRepairedDamage': 'unrepaired_damage', 
'dateCreated': 'ad_created', 
'nrOfPictures': 'nr_of_pictures', 
'postalCode': 'postal_code',
'lastSeen': 'last_seen'}
autos.rename(columns=cols_dict, inplace=True)
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

In the cells above we reviewed the column headings. We then created a dictionary mapping current column headings to new headings we want to replace them with, which would make our data easier to understand. We assigned these back into the dataframe, enabling the following amendments:
- Changed ` yearOfRegistration` to `registration_year`
- Changed `monthOfRegistration` to `registration_month`
- Changed `notRepairedDamage` to `unrepaired_damage`
- Changed `dateCreated` to `ad_created`
- Changed all other columns with spaces from camelcase to snakecase format

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

In [9]:
autos = autos.dropna(axis=1)

In [10]:
autos["price"] = autos["price"].astype(int)
autos["kilometer"] = autos["kilometer"].astype(int)
autos.rename({"kilometer": "odometer_km"}, axis=1, inplace=True)

In [11]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts())
sorted_km = autos["odometer_km"].sort_index(ascending=True)
print(sorted_km.head())
print(autos["odometer_km"].median())

(13,)
count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64
0    150000
1    125000
2    125000
3    150000
4     90000
Name: odometer_km, dtype: int32
150000.0


On investigating the `odometer_km` we can see the data appears to have a realistic distribution, and we don't need to exclude any outliers. Let's look at "price" now.

In [12]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().sort_index(ascending=True))

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64
0             10778
1              1189
2                12
3                 8
4                 1
              ...  
32545461          1
74185296          1
99000000          1
99999999         15
2147483647        1
Name: price, Length: 5597, dtype: int64


In [13]:
autos = autos[autos["price"].between(0,350000)]
print(autos["price"].value_counts())

0        10778
500       5670
1500      5394
1000      4649
1200      4594
         ...  
20470        1
25740        1
4994         1
8580         1
8188         1
Name: price, Length: 5545, dtype: int64


In [14]:
print(autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index())
print(autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index())
print(autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05 14:06:22    0.000003
2016-03-05 14:06:23    0.000003
2016-03-05 14:06:24    0.000008
2016-03-05 14:06:25    0.000005
2016-03-05 14:06:26    0.000003
                         ...   
2016-04-07 14:36:54    0.000003
2016-04-07 14:36:55    0.000003
2016-04-07 14:36:56    0.000005
2016-04-07 14:36:57    0.000003
2016-04-07 14:36:58    0.000003
Name: date_crawled, Length: 280446, dtype: float64
2014-03-10 00:00:00    0.000003
2015-03-20 00:00:00    0.000003
2015-06-11 00:00:00    0.000003
2015-06-18 00:00:00    0.000003
2015-08-07 00:00:00    0.000003
                         ...   
2016-04-03 00:00:00    0.038887
2016-04-04 00:00:00    0.037745
2016-04-05 00:00:00    0.011650
2016-04-06 00:00:00    0.003156
2016-04-07 00:00:00    0.001556
Name: ad_created, Length: 114, dtype: float64
2016-03-05 14:15:08    0.000003
2016-03-05 14:15:16    0.000003
2016-03-05 14:15:39    0.000003
2016-03-05 14:18:30    0.000003
2016-03-05 14:25:59    0.000003
                         ...   
2016-04

Above we have examined the `date_crawled`, `ad_created` and `last_seen` columns. 
We can see that `date_crawled` and `last_seen` columns correlate in terms of total of time (across all rows; a little over 1 month), and their date values. 

The `ad_created` column on the otherhand spans a much longer period (almost 2 years, though if we remove one outlier this reduces to 2 year). The last advert within this dataset was 7th April 2016. 

In [15]:
autos["registration_year"].describe()

count    371413.000000
mean       2004.561152
std          91.937676
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

We can see a few issues from the `registration_year` descriptive statistics:
- The earliest date is the year 1000
- The latest date is 9999
- We know that as the latest `ad_created` was in 2016, any value above this in `registration_year` is inaccurate.

We can remove data from this column, but before we do so let's look further at the earlier registration dates to see if we can identify at an appropriate cut-off point.

In [16]:
autos['registration_year'].value_counts(normalize=True).sort_index().head(20)

1000    0.000102
1001    0.000003
1039    0.000003
1111    0.000008
1200    0.000003
1234    0.000011
1253    0.000003
1255    0.000003
1300    0.000005
1400    0.000003
1500    0.000013
1600    0.000005
1602    0.000003
1688    0.000003
1800    0.000013
1910    0.000277
1911    0.000003
1915    0.000003
1919    0.000003
1920    0.000003
Name: registration_year, dtype: float64

In [17]:
autos['registration_year'].value_counts(normalize=True).sort_index().tail(50)

2010    0.033251
2011    0.032492
2012    0.025357
2013    0.016575
2014    0.012926
2015    0.008018
2016    0.026515
2017    0.028383
2018    0.010745
2019    0.000070
2066    0.000003
2200    0.000003
2222    0.000005
2290    0.000003
2500    0.000011
2800    0.000005
2900    0.000003
3000    0.000019
3200    0.000003
3500    0.000003
3700    0.000003
3800    0.000003
4000    0.000008
4100    0.000003
4500    0.000005
4800    0.000003
5000    0.000048
5300    0.000003
5555    0.000005
5600    0.000003
5900    0.000003
5911    0.000005
6000    0.000016
6200    0.000003
6500    0.000003
7000    0.000011
7100    0.000003
7500    0.000005
7777    0.000003
7800    0.000003
8000    0.000005
8200    0.000003
8455    0.000003
8500    0.000003
8888    0.000005
9000    0.000013
9229    0.000003
9450    0.000003
9996    0.000003
9999    0.000070
Name: registration_year, dtype: float64

We can see a jump from 1800 to 1910, the latter being a realistic registration year for a car - so we can remove any data in this column listed as lower than 1910.

In [18]:
autos = autos[autos["registration_year"].between(1910,2016)]
autos["registration_year"].describe()

count    356674.000000
mean       2002.781893
std           7.389662
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64

In [19]:
print(autos['brand'].value_counts(ascending=False))
print(autos['brand'].describe())
print(autos['brand'].value_counts(normalize=True,ascending=False).head(17))

volkswagen        75756
bmw               39126
opel              38197
mercedes_benz     34228
audi              31871
ford              24559
renault           16968
peugeot           10653
fiat               9189
seat               6647
skoda              5497
mazda              5475
smart              5032
citroen            4949
nissan             4841
toyota             4548
sonstige_autos     3775
hyundai            3507
mini               3286
volvo              3257
mitsubishi         2949
honda              2707
kia                2454
alfa_romeo         2264
suzuki             2253
porsche            2176
chevrolet          1788
chrysler           1406
dacia               874
daihatsu            780
jeep                779
land_rover          762
subaru              758
jaguar              613
trabant             577
saab                518
daewoo              513
rover               463
lancia              461
lada                218
Name: brand, dtype: int64
count         

When we look at the data within the 'brand' column we can observe the following:
- There are 40 brands listed
- Volkswagen is the most popular brand, with 75,756 adverts (21.24%)

Let's aggregate this column with the top 15 brands as the, and assign the mean price per brand as the value. 

In [20]:
mean_price_by_brand = {}
brands = ['volkswagen','bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'smart', 'citroen', 'nissan']
for b in brands:
    selected_rows = autos[autos['brand'] == b]
    mean_price = selected_rows['price'].mean()
    mean_price_by_brand[b] = mean_price

print(mean_price_by_brand)

{'volkswagen': 5231.3883124769, 'bmw': 8224.969380974288, 'opel': 2870.0537738565854, 'mercedes_benz': 8387.50280472128, 'audi': 8849.650309058392, 'ford': 3595.8644488782115, 'renault': 2366.3506011315417, 'peugeot': 3206.2663099596357, 'fiat': 2803.857220589836, 'seat': 4420.369038664059, 'skoda': 6460.418046207022, 'mazda': 3970.6332420091326, 'smart': 3586.280802861685, 'citroen': 3649.0161648817943, 'nissan': 4585.39599256352}


We can see from the mean price (dollars) by brand that Audi is the most expensive, averaging 8,849.65, closely followed by Mercedes-Benz and BMW at 8,387.50 and 8,224.97 respectively. Renault is the cheapest on average at 2,366.35, whilst Volkawagen and is around the middle. Let's narrow down our dictionary to the top 6 brands. 

In [21]:
mean_price_by_brand = {}
brands = ['volkswagen','bmw', 'opel', 'mercedes_benz', 'audi', 'ford']
for b in brands:
    selected_rows = autos[autos['brand'] == b]
    mean_price = selected_rows['price'].mean()
    mean_price_by_brand[b] = mean_price

print(mean_price_by_brand)

{'volkswagen': 5231.3883124769, 'bmw': 8224.969380974288, 'opel': 2870.0537738565854, 'mercedes_benz': 8387.50280472128, 'audi': 8849.650309058392, 'ford': 3595.8644488782115}


Let's compare mean price against mean mileage to see if we can identify any trends across the top brands.

In [22]:
avg_km_by_top_brands = {}
brands = ['volkswagen','bmw', 'opel', 'mercedes_benz', 'audi', 'ford']
for b in brands:
    selected_rows = autos[autos['brand'] == b]
    avg_km = selected_rows['odometer_km'].mean()
    avg_km_by_top_brands[b] = avg_km

print(avg_km_by_top_brands)

{'volkswagen': 128338.01943080415, 'bmw': 132666.64110821448, 'opel': 128756.83954237244, 'mercedes_benz': 130585.19340890499, 'audi': 129498.44686392018, 'ford': 123617.81831507798}


Now we can combine the two dictionaries into a new dataframe (using series constructor followed by dataframe constructor) to look at the average price against average odometer distance. 

In [23]:
bmp_series = pd.Series(mean_price_by_brand)

In [24]:
km_brand_series = pd.Series(avg_km_by_top_brands)

In [25]:
auto_brands = pd.DataFrame(bmp_series, columns=['mean_price'])
auto_brands

mean_price
volkswagen     5231.388312
bmw            8224.969381
opel           2870.053774
mercedes_benz  8387.502805
audi           8849.650309
ford           3595.864449

In [26]:
auto_brands["average_km"] = km_brand_series
auto_brands

mean_price     average_km
volkswagen     5231.388312  128338.019431
bmw            8224.969381  132666.641108
opel           2870.053774  128756.839542
mercedes_benz  8387.502805  130585.193409
audi           8849.650309  129498.446864
ford           3595.864449  123617.818315

Having analysed the top 6 brands' average price against their average odometer distance, we can see distance isn't significantly different between brands. However Opel, Ford and Volkswagen and substantially cheaper than the top 3 brands, which could potentially indicate better value for money. 

# Further Data Cleaning

There are additional steps we can take to clean the data. We know that some data refers to German words e.g. We'll start by identifying German words, and translating them into English. 

In [27]:
print(autos.head(),'\n')
print(autos['seller'].value_counts(ascending=False))
print('\n', autos['offer_type'].value_counts(ascending=False))


          date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest  registration_year  power_ps  odometer_km  registration_month  \
0    480   test               1993         0       150000                   0   
1  18300   test               2011       190       125000                   5   
2   9800   test               2004       163       125000                   8   
3   1500   test               2001        75       150000                   6   
4   3600   test               2008        69        90000                   7   

        brand           ad_created  nr_of_pictures

In [28]:
autos['seller'] = autos['seller'].str.replace("privat","public").str.replace("gewerblich","commercial")
autos['offer_type'] = autos['offer_type'].str.replace("Angebot","offer").str.replace("Gesuch","request")
print(autos['seller'].value_counts(), 
      '\n', 
      autos['offer_type'].value_counts())

public        356671
commercial         3
Name: seller, dtype: int64 
 offer      356662
request        12
Name: offer_type, dtype: int64
